plugins/certifier/certifier.cc

```diff
@@ -316,45 +316,28 @@ static TSMutex serial_mutex;     ///< serial number mutex
 static std::unique_ptr<SslLRUList> ssl_list = nullptr;
 static std::string store_path;
 
-/// Local helper function that generates a CSR based on common name
-static scoped_X509_REQ
-mkcsr(const char *cn)
+/// Local helper function that adds a Subject Alternative Name field into a
+/// certificate
+static void
+addSANExtToCert(X509 *cert, const std::string &dnsName)
 {
-  TSDebug(PLUGIN_NAME, "Entering mkcsr()...");
-  X509_NAME *n;
-  scoped_X509_REQ req;
-  req.reset(X509_REQ_new());
-
-  /// Set X509 version
-  X509_REQ_set_version(req.get(), 1);
-
-  /// Get handle to subject name
-  n = X509_REQ_get_subject_name(req.get());
-
-  /// Set common name field
-  if (X509_NAME_add_entry_by_txt(n, "CN", MBSTRING_ASC, (unsigned char *)cn, -1, -1, 0) != 1) {
-    TSError("[%s] mkcsr(): Failed to add entry.", PLUGIN_NAME);
-    return nullptr;
-  }
-  /// Set Traffic Server public key
-  if (X509_REQ_set_pubkey(req.get(), ca_pkey_scoped.get()) != 1) {
-    TSError("[%s] mkcsr(): Failed to set pubkey.", PLUGIN_NAME);
-    return nullptr;
-  }
-  /// Sign with Traffic Server private key
-  if (X509_REQ_sign(req.get(), ca_pkey_scoped.get(), EVP_sha256()) <= 0) {
-    TSError("[%s] mkcsr(): Failed to Sign.", PLUGIN_NAME);
-    return nullptr;
-  }
-  return req;
+  TSDebug(PLUGIN_NAME, "Adding SAN extension to the cert");
+  GENERAL_NAMES *generalNames = sk_GENERAL_NAME_new_null();
+  GENERAL_NAME *generalName   = GENERAL_NAME_new();
+  ASN1_IA5STRING *ia5         = ASN1_IA5STRING_new();
+  ASN1_STRING_set(ia5, dnsName.data(), dnsName.length());
+  // generalName owns ia5 after this call
+  GENERAL_NAME_set0_value(generalName, GEN_DNS, ia5);
+  sk_GENERAL_NAME_push(generalNames, generalName);
+  X509_add1_ext_i2d(cert, NID_subject_alt_name, generalNames, 0, X509V3_ADD_DEFAULT);
+  sk_GENERAL_NAME_pop_free(generalNames, GENERAL_NAME_free);
 }
 
-/// Local helper function that generates a X509 certificate based on CSR
+/// Local helper function that generates a X509 certificate based on SNI
 static scoped_X509
-mkcrt(X509_REQ *req, int serial)
+mkcrt(const std::string &commonName, int serial)
```

Comment:
Thanks for pointing out but I think in this function, `commonName` is passed into `X509_NAME_add_entry_by_txt()` which takes a `char*` so I would just use `const std::string &` since that's null-terminated.

Reply:
The std::string_view is null-terminated as well in this case, but it's nice to know it if you take std::string.